src/node_modules/app/elasticsearch/type-system.md

```diff
@@ -0,0 +1,226 @@
+	size: 'float', // defining a `size` request parameter if `float` type
+	count: 'integer', // defining a parameter of `float` type
+	name: 'string', // defining a parameter of 'string' type
+	date: 'Date_YYYYMMDD_dash', // Will present the user with a date editor
+	someParam: 'json' // the value will be parsed as json and added to the request
+}
+```
+
+#### String Literals
+
+String literal types are strings whose first and last characters are double
+quotes `"`. They're mainly used in conjunction with type unions which are
+explained immediately below.
+
+```javascript
+const aStringLiteral = '"homoscedastic"';
+```
+
+#### Type unions
+
+Sometimes a request parameter can have values of differing type. This can be
+specified in the type system with a type union, which is represented by an
+array of types.
+
+```javascript
+const t_test = {
+	// The following specifies a parameter allowing only three possible
+	// string values through a type union of three string literal types.
+	type: ['"paired"', '"homoscedastic"', '"heteroscedastic"']
+};
+```
+#### Compound types
+
+Compound types are defined using objects.
+
+```javascript
+// A compound type with two internal properties specifying type unions
+// of string literals.
+const sortOptions = {
+	_key: ['"asc"', '"desc"', 'undefined'],
+	_count: ['"asc"', '"desc"', 'undefined']
+};
+```
+
+#### Generic/dynamic types
+
+Generic types can be specified by providing a function whose arguments are
+the generic type arguments. At this time, generics is limited to a single
+generic type, which is all that is needed for Elasticsearch requests.
+
+```javascript
+// A generic or dynamic type defined through a function. The evaluated type
+// will depend on the value of the type.
+const metricFieldParams = type => ({
+	field: 'string',
+	missing: [
+		['integer', 'float'].includes(type)
+			? type
+			: 'string',
+		'undefined'
+	]
+});
+```
+
+#### Arrays
+
+We can specify array types in two ways:
+
+* Arrays of named types are strings ending with `[]`.
+* For types that can't be specified with single strings, arrays of these types
+  can also be defined with objects having a single `__array` property whose
+	value is the type of the array.
+
+```javascript
+	const percentile_ranks = {
+		// an array of floats
+		values: 'float[]', // could also be defined as `{__array: 'float'}`
+	};
+	const date_range = {
+		// an array of ranges
+		ranges: {
+			__array: {
+				key: ['string', 'undefined'],
+				from: ['string', 'undefined'],
+				to: ['string', 'undefined']
+			}
+		}
+	};
+```
+
+#### Numericals with units
+
+Numbers with units can be specified with an object having a single `__units`
+property whose values is an array of strings defining the possible units that 
+can be used.
+
+```javascript
+// A type specifying a physical quantity and its list of allowed units.
+const calendar_interval = {
+	__units: ['m', 'h', 'd', 'w', 'M', 'w', 'q', 'y']
+};
+```
+
+#### Record types
```

Comment:
as discussed this needs a simple example where we can see that keys are constrained